clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java

```diff
@@ -0,0 +1,224 @@
+  private final long mBufferBytes;
+  // Total number of bytes written to block, but not necessarily flushed to the remote server.
+  private long mWrittenBytes = 0;
+  // True if stream is open.
+  private boolean mOpen = false;
+
+  /**
+   * @param file the file the block belongs to
+   * @param opType the OutStream's write type
+   * @param blockIndex the index of the block in the file
+   * @param tachyonConf the TachyonConf instance for this file output stream.
+   * @throws IOException
+   */
+  RemoteBlockOutStream(TachyonFile file, WriteType opType, int blockIndex, TachyonConf tachyonConf)
+      throws IOException {
+    this(file, opType, blockIndex, tachyonConf.getBytes(Constants.USER_QUOTA_UNIT_BYTES,
+        8 * Constants.MB), tachyonConf);
+  }
+
+  /**
+   * @param file the file the block belongs to
+   * @param opType the OutStream's write type
+   * @param blockIndex the index of the block in the file
+   * @param initialBytes the initial size bytes that will be allocated to the block
+   * @param tachyonConf the TachyonConf instance for this file output stream.
+   * @throws IOException
+   */
+  RemoteBlockOutStream(TachyonFile file, WriteType opType, int blockIndex, long initialBytes,
+      TachyonConf tachyonConf) throws IOException {
+    super(file, opType, tachyonConf);
+
+    if (!opType.isCache()) {
+      throw new IOException("RemoteBlockOutStream only support WriteType.CACHE");
+    }
+
+    mBlockIndex = blockIndex;
+    mBlockCapacityBytes = mFile.getBlockSizeByte();
+    mBlockId = mFile.getBlockId(mBlockIndex);
+    mCloser = Closer.create();
+
+    // Create a local buffer.
+    mBufferBytes = mTachyonConf.getBytes(Constants.USER_FILE_BUFFER_BYTES, Constants.MB);
+    long allocateBytes = mBufferBytes;
+    mBuffer = ByteBuffer.allocate(Ints.checkedCast(allocateBytes));
+
+    // Open the remote writer.
+    mRemoteWriter =
+        mCloser.register(RemoteBlockWriter.Factory.createRemoteBlockWriter(tachyonConf));
+    mRemoteWriter.open(mTachyonFS.getWorkerDataServerAddress(), mBlockId, mTachyonFS.getUserId());
+    mOpen = true;
+  }
+
+  /**
+   * Write data to the remote block. This is synchronized to serialize writes to the block.
+   *
+   * @param bytes An array of bytes representing the source data.
+   * @param offset The offset into the source array of bytes.
+   * @param length The length of the data to write (in bytes).
+   * @throws IOException
+   */
+  private synchronized void writeToRemoteBlock(byte[] bytes, int offset, int length)
+      throws IOException {
+    mRemoteWriter.write(bytes, offset, length);
+  }
+
+  // Flush the local buffer to the remote block.
+  private void flushBuffer() throws IOException {
+    writeToRemoteBlock(mBuffer.array(), 0, mBuffer.position());
+    mBuffer.clear();
+  }
+
+  @Override
+  public void cancel() throws IOException {
+    if (mOpen) {
+      mCloser.close();
+      mOpen = false;
+      if (mWrittenBytes > 0) {
```

Comment:
Thanks for catching this! I added a variable keeping track of the flushed bytes.